emulator/bacnet-device-emulator.ts

```diff
@@ -0,0 +1,662 @@
+
+client.on('writeProperty', (data: any) => {
+	debug('writeProperty request', data)
+	try {
+		const payload = data.payload || {}
+		const sender = data.header?.sender
+
+		// Use incremental invokeId
+		const invokeId = data.invokeId ?? getNextInvokeId()
+		debug(`Using invokeId ${invokeId} for writeProperty response`)
+
+		const objectId = payload.objectId
+		const property = payload.property || payload.value?.property
+		const value = payload.value
+
+		if (!sender || !objectId || !property || value === undefined) {
+			debug('Missing required properties', {
+				sender,
+				invokeId,
+				objectId,
+				property,
+				value,
+			})
+			return
+		}
+
+		const objectKey = `${objectId.type}:${objectId.instance}`
+		const object = dataStore[objectKey]
+
+		if (!object) {
+			debug(`Object not found ${objectKey}, sending error response`)
+			client.errorResponse(
+				sender,
+				baEnum.ConfirmedServiceChoice.WRITE_PROPERTY,
+				invokeId,
+				baEnum.ErrorClass.OBJECT,
+				baEnum.ErrorCode.UNKNOWN_OBJECT,
+			)
+			debug(`Error response sent for unknown object`)
+			return
+		}
+
+		const propertyId = property.id
+		const propertyValue = object[propertyId]
+		if (!propertyValue) {
+			debug(`Property not found ${propertyId}, sending error response`)
+			client.errorResponse(
+				sender,
+				baEnum.ConfirmedServiceChoice.WRITE_PROPERTY,
+				invokeId,
+				baEnum.ErrorClass.PROPERTY,
+				baEnum.ErrorCode.UNKNOWN_PROPERTY,
+			)
+			debug(`Error response sent for unknown property`)
+			return
+		}
+
+		if (property.index === 0xffffffff) {
+			object[propertyId] = Array.isArray(value.value)
+				? value.value
+				: [value.value]
+			debug(
+				`Sending simpleAckResponse to ${typeof sender === 'string' ? sender : sender.address} for ${objectKey}:${propertyId} with invokeId ${invokeId}`,
+			)
+			client.simpleAckResponse(
+				sender,
+				baEnum.ConfirmedServiceChoice.WRITE_PROPERTY,
+				invokeId,
+			)
+			debug(`simpleAckResponse sent successfully`)
+		} else {
+			if (!propertyValue[property.index]) {
+				debug(
+					`Property index not found ${property.index}, sending error response`,
+				)
+				client.errorResponse(
+					sender,
+					baEnum.ConfirmedServiceChoice.WRITE_PROPERTY,
+					invokeId,
+					baEnum.ErrorClass.PROPERTY,
+					baEnum.ErrorCode.INVALID_ARRAY_INDEX,
+				)
+				debug(`Error response sent for invalid index`)
+				return
+			}
+
+			propertyValue[property.index] = Array.isArray(value)
+				? value[0]
+				: value
+			debug(
+				`Sending simpleAckResponse (with index) to ${typeof sender === 'string' ? sender : sender.address} for ${objectKey}:${propertyId}[${property.index}] with invokeId ${invokeId}`,
+			)
+			client.simpleAckResponse(
+				sender,
+				baEnum.ConfirmedServiceChoice.WRITE_PROPERTY,
+				invokeId,
+			)
+			debug(`simpleAckResponse (with index) sent successfully`)
+		}
+	} catch (error) {
+		debug('Error handling writeProperty request', error)
+	}
+})
+
+client.on('whoHas', (data: any) => {
```

Comment:
remove `any` from events types, our typed event emitter should type this correctly